web/static/js/lookit-jspsych.js

```diff
@@ -0,0 +1,121 @@
+        ?.split("=")[1];
+}
+
+async function get(url) {
+    /**
+     * Function for REST get.
+     */
+    const request = new Request(
+        url,
+        {
+            method: 'GET',
+            mode: 'same-origin'
+        }
+    );
+
+    const response = await fetch(request);
+    return await response.json();
+}
+
+
+async function patch(url, data) {
+    /**
+     * Function for REST patch.  
+     */
+    const request = new Request(
+        url,
+        {
+            method: 'PATCH',
+            headers: {
+                'X-CSRFToken': csrfToken(),
+                'Content-Type': 'application/vnd.api+json'
+            },
+            mode: 'same-origin', // Do not send CSRF token to another domain.
+            body: JSON.stringify({ data })
+        }
+    );
+
+    const response = await fetch(request);
+    return await response.json();
+}
+
+function on_data_update(responseApiUrl, id, userFunc) {
+    /**
+     * Function that returns a function to be used in place of jsPsych's option 
+     * "on_data_update".  "userFunc" should be the user's implementation of 
+     * "on_data_update".  Since this is the data that is returned from each 
+     * trial, this function will get the collected trial data and append the 
+     * current data point.
+     */
+    return async function (data) {
+        const { data: { attributes: { exp_data } } } = await get(responseApiUrl);
+
+        await patch(responseApiUrl, {
+            id, type: "responses", attributes: {
+                exp_data: [...exp_data, data]
+            }
+        });
+
+        // Don't call the function if not defined by user.
+        if (typeof userFunc === 'function') {
+            userFunc(data);
+        }
+    };
+}
+
+
+function on_finish(responseApiUrl, id, exitUrl, userFunc) {
+    /**
+     * Function that returns a function to be used in place of jsPsych's option 
+     * "on_finish".  "userFunc" should be the user's implementation of 
+     * "on_finish".  Since this is point where the experiment has ended, the 
+     * function will set "completed" to true and overwrites all experiment data 
+     * with the full set of collected data.  Once the user function has been 
+     * ran, this will redirect to the study's exit url.
+     */
+    return async function (data) {
+        await patch(responseApiUrl, {
+            id, type: "responses", attributes: {
+                exp_data: data.trials,
+                completed: true
+            }
+        });
+
+        // Don't call the function if not defined by user.
+        if (typeof userFunc === 'function') {
+            userFunc(data);
+        }
+
+        window.location.replace(exitUrl);
+    };
+}
+
+function lookitInitJsPsych(responseApiUrl, responseUuid, exitUrl) {
+    /**
+     * Function that returns a function to replace jsPsych's initJsPsych.
+     */
+    return function (opts) {
+        const jsPsych = origInitJsPsych({
+            ...opts,
```

Comment:
I've tried a few test experiments and I'm still getting an error here when I try to initialize jsPsych without passing an argument (which should be possible since none of the experiment setting arguments are required, and this is how it's done in the [hello world tutorial](https://www.jspsych.org/7.3/tutorials/hello-world/#step-3-create-a-script-element-and-initialize-jspsych)).
```js
const jsPsych = initJsPsych();
```
The error goes away as soon as I pass an argument into `initJsPsych`. 
So I think the function returned by `lookitInitJsPsych` just needs to check whether `opts` exists, or maybe use an empty object as a default:
```js
return function(opts={}) {
```

Reply:
This has been updated in a later commit.